### PR TITLE
[core] ROI tensor can set new ROI shape

### PR DESCRIPTION
**Pull request by praasz**
### Details:
 - ROI tensor allow set new shape of ROI. The initial ROI tensor shape is tensor's capacity. The new shape cane be set withing this capacity.

### Tickets:
 - [CVS-127815](https://jira.devtools.intel.com/browse/CVS-127815)

### CODE REVIEW

File: src/inference/src/dev/make_tensor.cpp [hunk 1]
```diff
@@ -297,7 +297,27 @@ class RoiTensor : public ITensor {
...
+        const auto tensor_rank = m_shape.size();
+        OPENVINO_ASSERT(new_shape.size() == tensor_rank);
+
+        auto owner_dim = m_owner->get_shape().cbegin();
+        auto stride = get_strides().cbegin();
+        auto new_dim = new_shape.cbegin();
+        auto dim = m_shape.cbegin();
+
+        for (size_t current_offset = m_offset; new_dim != new_shape.cend(); ++dim, ++new_dim, ++owner_dim, ++stride) {
+            const auto roi_coord = (current_offset / *stride);
```
Comment (ilya-lavrenov) [inline src/inference/src/dev/make_tensor.cpp:309]: do we need to take `m_offset` into account? 
maybe we can store initial `m_shape`  as `m_capacity` as well? and will check elementwise that `m_capacity` greater equal than `new_shape` ?

Comment (ilya-lavrenov) [reply src/inference/src/dev/make_tensor.cpp:309]: >  with current solution we can resize to the tensor shape.

Do you mean we can `set_shape` to be out of initial ROI? then it's prohibited.

Comment (praasz) [reply src/inference/src/dev/make_tensor.cpp:309]: Yes, ok so clear now and will apply changes.

**Review by ilya-lavrenov — APPROVED** [review]

File: src/core/tests/ov_tensor_test.cpp [hunk 2]
```diff
@@ -570,6 +570,50 @@ TEST_F(OVTensorTest, makeRangeRoiStringTensor) {
...
+    const ov::Shape new_shape({1, 2, 4, 3});
+    const ov::Shape roi_capacity({1, 2, 4, 3});
+
+    ASSERT_EQ(roi_tensor.get_shape(), roi_capacity);
+
+    roi_tensor.set_shape(new_shape);
+    ASSERT_EQ(roi_tensor.get_shape(), new_shape);
+
+    roi_tensor.set_shape(roi_capacity);
+    ASSERT_EQ(roi_tensor.get_shape(), roi_capacity);
```
Comment (t-jankowski) [inline src/core/tests/ov_tensor_test.cpp:596]: Values of `new_shape` and `roi_capacity` are equal. Shouldn't `new_shape` have smaller dimensions? It's not clear what is tested here.

File: src/inference/src/dev/make_tensor.cpp [hunk 3]
```diff
@@ -266,22 +289,15 @@ std::shared_ptr<ITensor> make_tensor(const element::Type element_type, const Sha
  */
 class RoiTensor : public ITensor {
 public:
-    RoiTensor(const std::shared_ptr<ITensor>& owner, const Coordinate& begin, const Coordinate& end) : m_owner{owner} {
+    RoiTensor(const std::shared_ptr<ITensor>& owner, const Coordinate& begin, const Coordinate& end)
+        : m_owner{owner},
```
Comment (t-jankowski) [inline src/inference/src/dev/make_tensor.cpp:293]: When taking ownership passing on `shared_ptr` by value is preferred.
```suggestion
    RoiTensor(std::shared_ptr<ITensor> owner, const Coordinate& begin, const Coordinate& end)
        : m_owner{std::move(owner)},
```

File: src/inference/src/dev/make_tensor.cpp [hunk 4]
```diff
@@ -266,22 +289,15 @@ std::shared_ptr<ITensor> make_tensor(const element::Type element_type, const Sha
...
+          m_shape{make_roi_shape(owner->get_shape(), begin, end)},
+          m_capacity{m_shape},
```
Comment (t-jankowski) [inline src/inference/src/dev/make_tensor.cpp:295]: Does any test cover the case when order of these members' definition changes?

File: src/core/tests/ov_tensor_test.cpp (hunk 2 above)
Comment (praasz) [reply src/core/tests/ov_tensor_test.cpp:596]: Shape should be diffrent will update.

**Review by t-jankowski — APPROVED** [review]

File: src/inference/src/dev/make_tensor.cpp (hunk 4 above)
Comment (praasz) [reply src/inference/src/dev/make_tensor.cpp:295]: if members oreder is worong there is compile warnning.

**Pull request merged**